File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/setup/GenerateTest.kt
package io.github.jkjamies.automationagent.agent.setup

import com.google.adk.kt.models.LlmRequest
import com.google.adk.kt.models.LlmResponse
import com.google.adk.kt.models.Model
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/** A deterministic [Model] that emits fixed text or throws. */
private class FixedModel(val text: String? = null, val error: Throwable? = null) : Model {
    override val name: String = "fixed"

    override fun generateContent(request: LlmRequest, stream: Boolean): Flow<LlmResponse> = flow {
        error?.let { throw it }
        emit(LlmResponse(content = assistantText(text ?: "")))
    }
}

class GenerateTest : BehaviorSpec({
    Given("a model that returns fixed text") {
        When("generating text") {
            Then("the concatenated response is returned") {
                generateText(FixedModel(text = "the answer"), "be terse", "question?") shouldBe "the answer"
            }
        }
    }

    Given("a model that errors") {
        When("generating text") {
            Then("the error propagates") {
                shouldThrow<RuntimeException> {
                    generateText(FixedModel(error = RuntimeException("model down")), "s", "u")
                }
            }
        }
    }
})

File: go/internal/agent/setup/generate_test.go
package setup

import (
	"context"
	"errors"
	"iter"
	"testing"

	"google.golang.org/adk/model"
)

type fixedLLM struct {
	text string
	err  error
}

func (f fixedLLM) Name() string { return "fixed" }
func (f fixedLLM) GenerateContent(context.Context, *model.LLMRequest, bool) iter.Seq2[*model.LLMResponse, error] {
	return func(yield func(*model.LLMResponse, error) bool) {
		if f.err != nil {
			yield(nil, f.err)
			return
		}
		yield(&model.LLMResponse{Content: AssistantText(f.text)}, nil)
	}
}

func TestGenerateText(t *testing.T) {
	out, err := GenerateText(context.Background(), fixedLLM{text: "the answer"}, "be terse", "question?")
	if err != nil {
		t.Fatalf("GenerateText: %v", err)
	}
	if out != "the answer" {
		t.Errorf("out = %q", out)
	}
}

func TestGenerateTextError(t *testing.T) {
	if _, err := GenerateText(context.Background(), fixedLLM{err: errors.New("model down")}, "s", "u"); err == nil {
		t.Fatal("expected error from model")
	}
}
